Integrate LRO Guidance to make Guidelines self-contained (#282)
Microsoft/api-guidelines: azure/LongRunningOperations.md

> # Long-Running Operations

Author: I made this a separate document for now -- I figured we could work out what the content should be and then decide where it should live.

Reviewer: This doc & content look good. One thing we may want to call out is that this is a REST binding to LROs and that implementations for other protocols should follow the same pattern. For example, in resource based pattern, you get a "shallow" representation of the resource. In stepwise, you get an alternate location to go ask for the state of the resource. 

Do we want to add anything about the identity of the resource in the two different models? For example, in resource oriented LROs, you can get back the shallow representation of the resource, which should have its ID. This would not be the case in stepwise.

What about idempotency? Should we explicitly call out that these are idempotent operations?

that holds the status of an outstanding or last completed operation.
This means that the client can use a standard "get" operation on the resource to determine the
status of an operation it initiated.  The flow looks like this:

![The RELO flow](./relo.jpg)

1. The client sends the initial request to the resource to initiate the long-running operation.
This initial request could be a PUT, PATCH, POST, or DELETE method.

2. The resource validates the request and initiates the operation processing.
It sends a response to client with a `200-OK` HTTP status code (or `201-Created` if the operation
is a create operation) and a representation of the resource where the `status` field is set
to a value indicating that the operation processing has been started.

3. The client then issues a GET request to the resource to determine if the operation processing
has completed.

4. The resource responds with a representation of the resource. While the operation is still being
processed, the status field will contain a "non-terminal" value, like "processing".

5. After the operation processing has completed, a GET request from the client will receive a response

> where the status field contains a "terminal" value -- `succeeded`, `failed`, or `cancelled` --

Reviewer 2: Inconsistent casing has been used "Failed" vs "failed".
Also, i think it's "Success"; not "succeeded".
Also, I think cancelled with 1 L and 2 Ls has also been used.
All this needs to be cleaned up.

Author: I did some research on "Success" vs "Succeeded":
- The [ARM guidelines](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#provisioningstate-property) say "Succeeded"
- Looking just at resource-manager API definitions in azure-rest-api-specs:
  - "Succeeded" appears as an enum value (well -- on a line by itself) 1339 times in 694 files
  - "Success" appears as an enum value 78 times in 68 files

I think it is also desirable for enum values to be consistent in their part of speech -- e.g. all verbs or all nouns.  "Success" is a noun.  I believe that "Succeeded", "Failed", and "Canceled" are all intransitive verbs.

So I think we want to use "Succeeded".

I will make the other corrections.

that indicates the result of the operation.

## Stepwise long-running operations

In the stepwise LRO pattern, the status and results of the operation are encapsulated into an
operation-results object that is distinct from the target resource and specific to the individual
operation request.  Here's what the stepwise LRO pattern looks like:

![The stepwise LRO flow](./stepwise.jpg)

1. The client sends the initial request to initiate the long-running operation.
As in the RELO pattern, the initial request could be a PUT, PATCH, POST, or DELETE method.

2. The resource validates the request and initiates the operation processing.
It sends a response to client with a `202-Accepted` HTTP status code.
Included in this response is an `Operation-location` response header with the absolute URL of an

> operation-results endpoint for this specific operation.

Reviewer 2: To me, this question comes up: Why does the client need the operation-location URL if the operation-result URL is well-known? I think we're over complicating.
Now, you might say because a POST to initiate the operation has to return the operation ID so it can be used to look up the operation in the operation-results collection but then why don't we say that operation should just be initated via a PUT to the operation-results collection and then the ID is passed by the client and so there is no need for the operation-location header at all. And, this makes starting the operation idempotent too.

Author: I think this confusion comes from my choice of naming.  I was calling the URL returned in the `Operation-location` header the "operation-results" URL.  Perhaps calling it the "operation-location" URL is better -- or maybe just "operation URL"?

The response also includes a `Retry-after` header telling the client a minimum time to wait (in seconds)

> before sending a request to the operation-results URI.

Reviewer 2: URI or URL - please be consistent

Author: Fixed.  There is now just one occurrence of "URI" and that is within the name of `414` status code, which is the official name so I kept it.

3. After waiting at least the amount of time specified by the previous response's `Retry-after` header,
the client issues a GET request to the operation-results URI.

> 4. The operation-results URI responds with an operation-results object, containing information about the

Reviewer 2: ```suggestion
4. The operation-results URL responds with an operation-results resource containing information about the
```

operation including its current status. If the operation is still being processed,
the status field will contain a "non-terminal" value, like "processing".

> 5. After the operation processing has completed, a GET request to operation-results URI will return an

Reviewer 2: ```suggestion
5. After the operation processing completes, a GET request to operation-results URL returns an
```

> operation-results object with a status field containing a "terminal" value -- `succeeded`, `failed`, or `cancelled` --

Reviewer 2: ```suggestion
operation-results resource with its status field containing a terminal value: `succeeded`, `failed`, or `cancelled` --
```

Reviewer 2: Fix values for consistency

that indicates the result of the operation.

> If the status is `failed`, the operation-results object will contain an `error` field with a `code` and `message`

Reviewer 2: ```suggestion
If the status is `failed`, the operation-results resource must contain an `error` field with a `code` and `message`
```

Reviewer 2: We really need an example here
